crates/edr_napi/src/trace/message_trace.rs

```diff
@@ -0,0 +1,177 @@
+    pub bytecode: Option<Value>,
+    pub number_of_subtraces: u32,
+    // `CreateMessageTrace`
+    // NOTE: we can't use Option<T> as this is converted to an optional property,
+    // which is not backwards compatible with the current JS interface
+    pub deployed_contract: Either<Uint8Array, Undefined>,
+}
+
+#[napi(object)]
+pub struct CallMessageTrace {
+    // `BaseMessageTrace`
+    pub value: BigInt,
+    pub return_data: Uint8Array,
+    pub exit: ClassInstance<Exit>,
+    pub gas_used: BigInt,
+    pub depth: u32,
+    // `BaseEvmMessageTrace`
+    pub code: Uint8Array,
+    pub steps: Vec<Either4<EvmStep, PrecompileMessageTrace, CreateMessageTrace, CallMessageTrace>>,
+    // TODO: Will be later filled on the JS side but we should port to ContractsIdentifier in Rust
+    // This is explicitly `any` on the JS side to side-step the type-checking until we port
+    pub bytecode: Option<Value>,
+    pub number_of_subtraces: u32,
+    // `CallMessageTrace`
+    pub calldata: Uint8Array,
+    pub address: Uint8Array,
+    pub code_address: Uint8Array,
+}
+
+/// Converts [`edr_solidity::message_trace::MessageTraceStep`] to the N-API
+/// representation.
+///
+/// # Panics
+/// This function will panic if the value is mutably borrowed.
+pub fn message_trace_step_to_napi(
+    value: edr_solidity::message_trace::MessageTraceStep,
+    env: Env,
+) -> Either4<EvmStep, PrecompileMessageTrace, CreateMessageTrace, CallMessageTrace> {
+    match value {
+        edr_solidity::message_trace::MessageTraceStep::Evm(step) => {
+            Either4::A(EvmStep { pc: step.pc as u32 })
+        }
+        edr_solidity::message_trace::MessageTraceStep::Message(msg) => {
+            // Immediately drop the borrow lock as it may be
+            let owned = msg.borrow().clone();
+            match message_trace_to_napi(owned, env) {
+                Either3::A(precompile) => Either4::B(precompile),
+                Either3::B(create) => Either4::C(create),
+                Either3::C(call) => Either4::D(call),
+            }
+        }
+    }
+}
+
+/// Converts the Rust representation of a `MessageTrace` to the N-API
+/// representation.
+pub fn message_trace_to_napi(
+    value: edr_solidity::message_trace::MessageTrace,
+    env: Env,
+) -> Either3<PrecompileMessageTrace, CreateMessageTrace, CallMessageTrace> {
+    match value {
+        edr_solidity::message_trace::MessageTrace::Precompile(precompile) => {
+            Either3::A(PrecompileMessageTrace {
+                value: BigInt {
+                    sign_bit: false,
+                    words: precompile.base.value.as_limbs().to_vec(),
+                },
+                return_data: Uint8Array::from(precompile.base.return_data.as_ref()),
+                exit: Exit(precompile.base.exit.into())
+                    .into_instance(env)
+                    .expect("napi glue to register the Exit class constructor"),
```

Comment:
I know and was considering it but I decided against it, because IMO it'd obscure a bit what's actually happening here and the error can only occur if there's an internal bug in the napi-rs itself. If I recall correctly, the error can only happen if napi failed to successfully register the class constructors at init time; if that happens then we probably can't reason much about the glue code anymore.

On the flip side, I see that we're not using `#[napi(catch_unwind)]` so I'm happy to change at least so that we don't bring down the entire runtime if the napi glue code fails.